esmvalcore/_citation.py

```diff
@@ -0,0 +1,208 @@
+"""Citation module."""
+import os
+import logging
+import re
+from pathlib import Path
+import requests
+
+from ._config import DIAGNOSTICS_PATH
+
+if DIAGNOSTICS_PATH:
+    REFERENCES_PATH = Path(DIAGNOSTICS_PATH) / 'references'
+else:
+    REFERENCES_PATH = ''
+
+logger = logging.getLogger(__name__)
+
+CMIP6_URL_STEM = 'https://cera-www.dkrz.de/WDCC/ui/cerasearch'
+
+# it is the technical overview and should always be cited
+ESMVALTOOL_PAPER_TAG = 'righi19gmdd'
+ESMVALTOOL_PAPER = (
+    '@article{righi19gmdd,\n\t'
+    'doi = {10.5194/gmd-2019-226},\n\t'
+    'url = {https://doi.org/10.5194%2Fgmd-2019-226},\n\t'
+    'year = 2019,\n\t'
+    'month = {sep},\n\t'
+    'publisher = {Copernicus {GmbH}},\n\t'
+    'author = {Mattia Righi and Bouwe Andela and Veronika Eyring '
+    'and Axel Lauer and Valeriu Predoi and Manuel Schlund '
+    'and Javier Vegas-Regidor and Lisa Bock and Björn Brötz '
+    'and Lee de Mora and Faruk Diblen and Laura Dreyer '
+    'and Niels Drost and Paul Earnshaw and Birgit Hassler '
+    'and Nikolay Koldunov and Bill Little and Saskia Loosveldt Tomas '
+    'and Klaus Zimmermann},\n\t'
+    'title = {{ESMValTool} v2.0 '
+    '{\\&}amp$\\mathsemicolon${\\#}8211$\\mathsemicolon$ '
+    'Technical overview}\n'
+    '}\n'
+)
+
+
+def _write_citation_file(filename, provenance):
+    """
+    Write citation information provided by the recorded provenance.
+
+    Recipe and cmip6 data references are saved into one bibtex file.
+    cmip6 data references are provided by CMIP6 data citation service.
+    each cmip6 data reference has a json link. In the case of internet
+    connection, cmip6 data references are saved into a bibtex file.
+    Otherwise, cmip6 data reference links are saved into a text file.
+    """
+    # collect info from provenance
+    product_name = os.path.splitext(filename)[0]
+    info_urls = []
+    json_urls = []
+    product_tags = []
+    for item in provenance.records:
+        for key, value in item.attributes:
+            if key.namespace.prefix == 'attribute':
+                if key.localpart in {'reference', 'references'}:
+                    product_tags.append(value)
+                elif key.localpart == 'mip_era' and value == 'CMIP6':
+                    url_prefix = _make_url_prefix(item.attributes)
+                    info_urls.append(_make_info_url(url_prefix))
+                    json_urls.append(_make_json_url(url_prefix))
+
+    _save_citation_info(product_name, product_tags, json_urls, info_urls)
+
+
+def _save_citation_info(product_name, product_tags, json_urls, info_urls):
+    citation_entries = [ESMVALTOOL_PAPER]
+    citation_urls = ''
+
+    # save CMIP6 url_info, if any
+    if info_urls:
+        for info_url in info_urls:
+            citation_urls += '{}\n'.format(info_url)
+        with open(f'{product_name}_data_citation_url.txt', 'w') as file:
+            file.write(citation_urls)
+
+    # convert json_urls to bibtex entries
+    if json_urls:
```

Comment:
there is no need for this `if` statement, the loop below will do nothing if json_urls is an empty list.

Reply:
@bouweandela , thanks, I applied your suggestion.